run.py

```diff
@@ -1,116 +1,441 @@
+        monmap_loc = self._find_correct_monmap(self.machines)
+        if monmap_loc is None:
+            print 'No Mon has correct monmap, recovery impossible, aborting'
+            return
 
-    def _restart_ceph_services(self, addr, username, password):
-        connection = self._get_connection(addr, username, password)
-        self._execute_command(connection, 'sudo start ceph-all')
-        print addr + 'Restart successful'
+        self._inject_mon_map(monmap_loc, self.machines)
+
+        # Let's wait for sometime before getting ceph health status
+        time.sleep(10)
+        try:
+            self.check_ceph_cli_health(self.connection)
+        except TimeoutError:
+            print "Injecting Monmap didn't work, probably network issue"
+
+    def _correct_skew(self, skew_list):
+        for mon in skew_list:
+            for machine in self.machines:
+                if mon == machine.mon_id:
+                    if self.init_type == 'systemd':
+                        cmd = 'sudo systemctl restart ntp.service'
+                    else:
+                        cmd = 'sudo service ntp restart'
+                    out, err = self._execute_command(machine.connection, cmd)
+                    try:
+                        self._get_eof(out, cmd)
+                    except TimeoutError:
+                        print "Couldn't restart ntp for: ", mon
+                    else:
+                        print 'Restart successful for: ', mon
+
+    def _detect_clock_skew(self, connection):
+        out, err = self._execute_command(connection, 'sudo ceph health')
+
+        try:
+            self._get_eof(out, 'sudo ceph health')
+        except TimeoutError:
+            raise TimeoutError('sudo ceph health timed out')
+
+        status = out.read().split(' ', 1)
+
+        if not (len(status) == 2 and status[0].strip() in self.BAD_HEALTH):
+            return None
+
+        if re.search(r'clock skew detected on.*', status[1]) is None:
+            return None
+
+        skew_list = [i.strip()[4:] for i in
+                     re.findall(r'(mon\..*)[;]', status[1])[0].split(',')]
+        return skew_list
+
+    def _inject_mon_map(self, monmap_loc, machine_list):
+        for machine in machine_list:
+            if (machine.ssh_status == 'LIVE' and
+               machine.is_monmap_correct is False):
+                print 'Injecting monmap to: ' + machine.host
+
+                self._restart_ceph_mon_service('stop', machine.host)
+
+                machine.connection.open_sftp().put(monmap_loc,
+                                                   '/tmp/monmap')
+                cmd = 'sudo ceph-mon -i ' + machine.mon_id +\
+                    ' --inject-monmap /tmp/monmap'
+
+                self._execute_command(machine.connection, cmd)
+
+                self._restart_ceph_mon_service('start', machine.host)
+
+    def _find_correct_monmap(self, machine_list):
```

Comment:
This is a tricky function to write.  You probably want to do either:
1. Find the monmap for the monitor that is the leader of the quorum
2. Find the monmap that has the highest epoch